File: src/main/java/santiagoAndFerdy/vgs/discovery/selector/WeighedRandomSelector.java
package santiagoAndFerdy.vgs.discovery.selector;

import java.util.Map;
import java.util.Optional;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

/**
 * Created by Fydio on 4/5/16.
 */
public class WeighedRandomSelector implements ISelector {
    public static WeighedRandomSelector instance = new WeighedRandomSelector();
    public ThreadLocalRandom rng;

    public WeighedRandomSelector() {
        rng = ThreadLocalRandom.current();
    }

    @Override
    public Optional<Integer> selectIndex(Map<Integer, Long> weights) {
        long n = weights.values().stream().reduce(0L, (a, b) -> a + b);
        n += weights.size();

        Random rng = new Random();
        long random = -1;

        // keep generating a random number until a suitable one is found.
        while (random == -1) {
            long randomTry = new Double(Math.floor(rng.nextDouble() * n)).longValue();
            if(randomTry <= n) random = randomTry;
        }

        int counter = 0;
        Optional<Integer> i = Optional.empty();

        for (int key : weights.keySet()) {
            long weight = weights.get(key);

            if(counter <= random && (counter + weight + 1) > random) {
                i = Optional.of(key);
                break;
            };

            counter += weight + 1;
        }

        return i;
    }
}

File: src/main/scala/grid/discovery/Selector/WeighedRandomSelector.scala
package grid.discovery.Selector

import scala.util.Random
import scala.util.control.Breaks._

object WeighedRandomSelector extends Selector {
  val rng: Random = new Random()

  override def selectIndex(weights: Map[Int, Long]): Option[Int] = {
    if(weights.isEmpty) {
      None
    } else {

      val sum = weights.values.sum + weights.size

      var randomResult: Long = -1
      while (randomResult == -1) {
        import math._
        val randomTry = floor(rng.nextDouble() * sum).toLong

        if (randomTry <= sum) randomResult = randomTry
      }

      var index: Option[Int] = None
      var counter = 0L
      val keys = weights.keys.toList
      var cursor = 0

      breakable {
        while(cursor < keys.length) {
          if (counter <= randomResult && (counter + weights(keys(cursor)) + 1) > randomResult) {
            index = Some(keys(cursor))
            break
          }
          counter += weights(keys(cursor)) + 1
          cursor += 1
        }
      }

      index
    }
  }
}
